api/filtering/custom_filters.py

```diff
@@ -5,11 +5,32 @@
 logger = get_logger(__name__)
 
 
+def _build_no_minor_filter(operation, major, name):
+    return {"spf_operating_system": {"major": {operation: major}, "name": {"eq": name}}}
+
+
+def _build_major_equality_filter(major, minor, name, operation):
+    return {"spf_operating_system": {"major": {"eq": major}, "minor": {operation: minor}, "name": {"eq": name}}}
+
+
+# build filter based on the operation and OS name and version
 def _build_operating_system_version_filter(major, minor, name, operation):
-    os_filter = {"spf_operating_system": {"major": {"eq": major}, "minor": {operation: minor}, "name": {"eq": name}}}
+    if operation == "eq":
+        if minor is None:
+            os_filter = _build_no_minor_filter(operation, major, name)
+        else:
+            os_filter = _build_major_equality_filter(major, minor, name, operation)
+
+    if operation in ["lt", "gt", "lte", "gte"]:
+        if minor is None:
+            os_filter = _build_no_minor_filter(operation, major, name)
+        else:
+            # handle 'eq' here and 'gt' 'lt' in !='eq'
+            os_filter = _build_major_equality_filter(major, minor, name, operation)
```

Comment:
The branching on operation type is unnecessary, both branches do the same thing. We can just short circuit if the minor version is missing and leave the rest of the logic the same. I chose to inline the function calls since they are only used in one place.
```suggestion
    if minor is None:
        return {"spf_operating_system": {"major": {operation: major}, "name": {"eq": name}}}

    os_filter = {"spf_operating_system": {"major": {"eq": major}, "minor": {operation: minor}, "name": {"eq": name}}}
```

Reply:
I knew this would be questioned immediately but left it in for clarity because it explicitly shows how each operator is getting filtered; makes the code easy to read for someone not too familiar.  That said, the code is update per the recommendation here.  The code compact but hides the filter building complexity.  IMO, the explicit code is easier to follow, but the change is here if you don't buy the explicit reasoning.

Reply:
Personally, I think the longer and more explicit version is harder to read and understand, because it creates a logical split where both branches do the same thing

Reply:
Ok